tensorflow_probability/python/experimental/vi/cascading_flows.py

```diff
@@ -0,0 +1,607 @@
+
+    """
+    with tf.name_scope(name or 'build_cf_surrogate_posterior'):
+        surrogate_posterior, variables = _cf_surrogate_for_distribution(
+            dist=prior,
+            base_distribution_surrogate_fn=functools.partial(
+                _cf_convex_update_for_base_distribution,
+                initial_prior_weight=initial_prior_weight,
+                num_auxiliary_variables=num_auxiliary_variables,
+                num_layers=num_layers),
+            num_auxiliary_variables=num_auxiliary_variables,
+            num_layers=num_layers,
+            seed=seed)
+        surrogate_posterior.also_track = variables
+        return surrogate_posterior
+
+
+def _cf_surrogate_for_distribution(dist,
+                                   base_distribution_surrogate_fn,
+                                   num_auxiliary_variables,
+                                   num_layers,
+                                   global_auxiliary_variables=None,
+                                   sample_shape=None,
+                                   variables=None,
+                                   seed=None):
+    """Recursively creates CF surrogates, and creates new variables if needed.
+
+    Args:
+      dist: a `tfd.Distribution` instance.
+      base_distribution_surrogate_fn: Callable to build a surrogate posterior
+        for a 'base' (non-meta and non-joint) distribution, with signature
+        `surrogate_posterior, variables = base_distribution_fn(
+        dist, sample_shape=None, variables=None, seed=None)`.
+      num_auxiliary_variables: The number of auxiliary variables to use for each
+        variable in the input program.
+      num_layers: Number of layers to use in each Highway Flow architecture.
+      global_auxiliary_variables: The sampled global auxiliary variables
+        (available only if using auxiliary variables). Default value: None.
+      sample_shape: Optional `Tensor` shape of samples drawn from `dist` by
+        `tfd.Sample` wrappers. If not `None`, the surrogate's event will include
+        independent sample dimensions, i.e., it will have event shape
+        `concat([sample_shape, dist.event_shape], axis=0)`.
+        Default value: `None`.
+      variables: Optional nested structure of `tf.Variable`s returned from a
+        previous call to `_cf_surrogate_for_distribution`. If `None`,
+        new variables will be created; otherwise, constructs a surrogate posterior
+        backed by the passed-in variables.
+        Default value: `None`.
+      seed: Python `int` seed for random initialization.
+    Returns:
+      surrogate_posterior: Instance of `tfd.Distribution` representing a trainable
+        surrogate posterior distribution, with the same structure and `name` as
+        `dist`, and with addition of global and local auxiliary variables if
+        `num_auxiliary_variables > 0`.
+      variables: Nested structure of `tf.Variable` trainable parameters for the
```

Comment:
Since the components of `variables` are now bijectors, maybe we should call it a nested structure 'containing' `tf.Variables` rather than 'of' `tf.Variables` ?